### DISCUSSION TITLE
import ADS1x15 results in errors

### DISCUSSION THREAD

**Discussion by Mr-bkmtech** [micropython/micropython#11539] [RP2040 / Pico]
I am trying to communicate with an ADS1015 module using I2C and I get the following errors using Thonny as my editor

This is line 98:
import ADS1x15

I get the following messages:
Traceback (most recent call last):
  File "<stdin>", line 98, in <module>
  File "/lib/ADS1x15/__init__.py", line 5, in <module>
  File "/lib/ADS1x15/ADS1x15.py", line 1, in <module>
  File "/lib/smbus2/__init__.py", line 23, in <module>
  File "/lib/smbus2/smbus2.py", line 25, in <module>
  File "/lib/fcntl.py", line 1, in <module>
ImportError: no module named 'ffi'

I have gone through the Thonny packages available and cannot find a package that works, or even an easily identifiable package that should work.

I am not sure how to proceed at this point!

**Comment by jimmo** [comment]
It sounds like you're using a driver from PyPI that isn't designed for MicroPython (i.e. it's designed for a Raspberry Pi or some other Linux-based embedded system).

(As far as I understand it, Thonny will show _all_ PyPI packages, not just MicroPython-suitable ones).

**Reply by Mr-bkmtech** [reply]
That is correct

(As far as I understand it, Thonny will show all PyPI packages, not just MicroPython-suitable ones).

**Comment by mcauser** [comment]
Try this one: https://github.com/robert-hh/ads1x15

**Reply by Mr-bkmtech** [reply]
Hello

Thank you for the info.

Bear in mind I am relatively new to both micro python and Thonny but I am unsure how to integrate this code into my program.

I can run the code on the pico but it returns nothing.

**Reply by robert-hh** [reply]
Can you show us the code that you used to run the ADC?

**Reply by Mr-bkmtech** [reply]
Let me give you this as this is the code that runs successfully on a Pi3 and Pi0

As I understand it this was adapted from Arduino code, and it runs on everything but the Pi Pico

```
#!/usr/bin/python3
# $Id: simpletest,v 1.1 2021/02/01 23:13:59 heegard Exp heegard $

# Simple demo of reading each analog input from the ADS1x15 and printing it to
# the screen.
# Author: Tony DiCola
# License: Public Domain
import time

# Import the ADS1x15 module.
import Adafruit_ADS1x15


# Create an ADS1115 ADC (16-bit) instance.
#adc = Adafruit_ADS1x15.ADS1115()

# Or create an ADS1015 ADC (12-bit) instance.
adc = Adafruit_ADS1x15.ADS1015()

# Note you can change the I2C address from its default (0x48), and/or the I2C
# bus by passing in these optional parameters:
#adc = Adafruit_ADS1x15.ADS1015(address=0x49, busnum=1)

# Choose a gain of 1 for reading voltages from 0 to 4.09V.
# Or pick a different gain to change the range of voltages that are read:
#  - 2/3 = +/-6.144V
#  -   1 = +/-4.096V
#  -   2 = +/-2.048V
#  -   4 = +/-1.024V
#  -   8 = +/-0.512V
#  -  16 = +/-0.256V
# See table 3 in the ADS1015/ADS1115 datasheet for more info on gain.
GAIN = 8

print('Reading ADS1x15 values, press Ctrl-C to quit...')
# Print nice channel column headers.
print('| {0:>6} | {1:>6} | {2:>6} | {3:>6} |'.format(*range(4)))
print('-' * 37)
# Main loop.
while True:
    # Read all the ADC channel values in a list.
    values = [0]*4
    for i in range(4):
        # Read the specified ADC channel using the previously set gain value.
        values[i] = adc.read_adc(i, gain=GAIN)
        # Note you can also pass in an optional data_rate parameter that controls
        # the ADC conversion time (in samples/second). Each chip has a different
        # set of allowed data rate values, see datasheet Table 9 config register
        # DR bit values.
        #values[i] = adc.read_adc(i, gain=GAIN, data_rate=128)
        # Each value will be a 12 or 16 bit signed integer value depending on the
        # ADC (ADS1015 = 12-bit, ADS1115 = 16-bit).
    # Print the ADC values.
    print('| {0:>6} | {1:>6} | {2:>6} | {3:>6} |'.format(*values))
    # Pause for half a second.
    time.sleep(0.5)
```

**Reply by robert-hh** [reply]
Please enclose Python code in lines with three backticks ``` to make it readable.

**Reply by Mr-bkmtech** [reply]
Hi

Done

Sorry about that

**Reply by robert-hh** [reply]
No problem. As you state, the driver you tried is not compatible to MicroPython. So you cannot use it. You can use the driver I adapted in 2017, located here: https://github.com/robert-hh/ads1x15. More MicroPython drivers may be available at https://awesome-micropython.com/, but they all point back to same same source.

**Reply by Mr-bkmtech** [reply]
Hi Robert

I tried this code but it does nothing for me.

Using Thonny, I run the code and this is what I get

>>> %Run -c $EDITOR_CONTENT
>>> 

Your code may work but I don't understand how to use it.

**Reply by robert-hh** [reply]
Just a simple example for RP2, reading channel 0 and print the ADC reading.
```
# Sample code for RP2. Read a single channel

from machine import I2C, Pin
from ads1x15 import ADS1015
i2c=I2C(0, sda=Pin(4), scl=Pin(5))
adc = ADS1015(i2c, address=72, gain=1)
value = adc.read(0, 0)
print(value)
```

**Reply by Mr-bkmtech** [reply]
Hi Robert

I appreciate your help.

Doesn't ads1x15 have to be added as a library first?

**Reply by robert-hh** [reply]
Of course you have to move the file ads1x15.py to the target board.

**Reply by Mr-bkmtech** [reply]
Hi Robert

Could you please describe the process because I am missing something.

Maybe I am going about this the wrong way because I am trying to learn 2 things, both Micro Python and Thonny.

**Reply by robert-hh** [reply]
You can use mpremote to copy files to your board. mpremote is the official tool by MicroPython, and can be installed using upip. Using Thonny that should be possible as well. But I cannot tell you anything about Thonny.

**Reply by rkompass** [reply]
Thonny has a menu entry like 'View'. There you may check something like 'Files' to get an extra view of the files on your computer and, as soon as you connected to the RP2, of the files on the RP2 file system. You may copy files by right-clicking on them and selecting 'copy ...'. Now you have to navigate in the upper view of files on your PC to ads1x15.py and then use the right mouse click to copy it to your RP2. Note that you can select different directories on the RP2 too. Files in the base directory '/' as well as in '/lib' should be found in an attempt to import them.
I assume that you may connect to the RP2 with Thonny (you have to select the Interpreter 'MicrPython (Raspberry Pi Pico)' under the menu entry 'Run' - 'select Interpreter' and after that perhaps several times touch the red 'Stop' icon).

**Reply by Mr-bkmtech** [reply]
Thanks

I got it working

**Reply by Mr-bkmtech** [reply]
> You can use mpremote to copy files to your board. mpremote is the official tool by MicroPython, and can be installed using upip. Using Thonny that should be possible as well. But I cannot tell you anything about Thonny.

I got it working, thank you

**Comment by Mr-bkmtech** [comment]
I will review all replies with my associate, thank you
